test/server/listener_manager_impl_test.cc

```diff
@@ -0,0 +1,242 @@
+        "config" : {}
+      }
+    ]
+  }
+  )EOF";
+
+  Json::ObjectSharedPtr loader = Json::Factory::loadFromString(json);
+  EXPECT_THROW_WITH_MESSAGE(manager_.addListener(*loader), EnvoyException,
+                            "unable to create filter factory for 'echo'/'write'");
+}
+
+class TestStatsConfigFactory : public Configuration::NamedNetworkFilterConfigFactory {
+public:
+  // Server::Configuration::NamedNetworkFilterConfigFactory
+  Configuration::NetworkFilterFactoryCb
+  createFilterFactory(const Json::Object&, Configuration::FactoryContext& context) override {
+    context.scope().counter("bar").inc();
+    return [](Network::FilterManager&) -> void {};
+  }
+  std::string name() override { return "stats_test"; }
+  Configuration::NetworkFilterType type() override {
+    return Configuration::NetworkFilterType::Read;
+  }
+};
+
+TEST_F(ListenerManagerImplTest, StatsScopeTest) {
+  Registry::RegisterFactory<TestStatsConfigFactory, Configuration::NamedNetworkFilterConfigFactory>
+      registered;
+
+  std::string json = R"EOF(
+  {
+    "address": "tcp://127.0.0.1:1234",
+    "bind_to_port": false,
+    "filters": [
+      {
+        "type" : "read",
+        "name" : "stats_test",
+        "config" : {}
+      }
+    ]
+  }
+  )EOF";
+
+  Json::ObjectSharedPtr loader = Json::Factory::loadFromString(json);
+  EXPECT_CALL(factory_, create_(_, false));
+  manager_.addListener(*loader);
+  manager_.listeners().front().get().listenerScope().counter("foo").inc();
+
+  EXPECT_EQ(1UL, server_.stats_store_.counter("bar").value());
+  EXPECT_EQ(1UL, server_.stats_store_.counter("listener.127.0.0.1_1234.foo").value());
+}
+
+/**
+ * Config registration for the echo filter using the deprecated registration class.
+ */
+class TestDeprecatedEchoConfigFactory : public Configuration::NetworkFilterConfigFactory {
+public:
+  // NetworkFilterConfigFactory
+  Configuration::NetworkFilterFactoryCb
+  tryCreateFilterFactory(Configuration::NetworkFilterType type, const std::string& name,
+                         const Json::Object&, Server::Instance&) override {
+    if (type != Configuration::NetworkFilterType::Read || name != "echo_deprecated") {
+      return nullptr;
+    }
+
+    return [](Network::FilterManager&) -> void {};
+  }
+};
+
+TEST_F(ListenerManagerImplTest, DeprecatedFilterConfigFactoryRegistrationTest) {
+  // Test ensures that the deprecated network filter registration still works without error.
+
+  // Register the config factory
+  Configuration::RegisterNetworkFilterConfigFactory<TestDeprecatedEchoConfigFactory> registered;
+
+  std::string json = R"EOF(
+  {
+    "address": "tcp://127.0.0.1:1234",
+    "filters": [
+      {
+        "type" : "read",
+        "name" : "echo_deprecated",
+        "config" : {}
+      }
+    ]
+  }
+  )EOF";
+
+  Json::ObjectSharedPtr loader = Json::Factory::loadFromString(json);
+  EXPECT_CALL(factory_, create_(_, true));
+  manager_.addListener(*loader);
+}
+
+} // Server
+} // Envoy
```

Comment:
Logically, it's all code movement. There are changes to the tests since now they just operate on a single listener vs. an entire config, however this makes the tests simpler.